lib/model.js

```diff
@@ -3499,6 +3499,46 @@ Model.bulkWrite = function(ops, options, callback) {
   }, this.events);
 };
 
+/**
+ *  takes an array of documents, gets the changes and inserts/updates documents in the database
+ *  according to whether or not the document is new, or whether it has changes or not.
+ *
+ * `bulkSave` uses `bulkWrite` under the hood, so it's mostly useful when dealing with many documents (10K+)
+ *
+ * @param {[Document]} documents
+ *
+ */
+Model.bulkSave = function(documents) {
+  if (!Array.isArray(documents)) {
+    throw new Error(`bulkSave expects an array of documents to be passed, received \`${documents}\` instead`);
+  }
+
+
+  const writeOperations = documents.reduce((accumulator, document, i) => {
+    if (!(document instanceof Document)) {
+      throw new Error(`documents.${i} was not a mongoose document, documents must be an array of mongoose documents (instanceof mongoose.Document).`);
+    }
+
+    const changes = document.getChanges();
+    if (document.isNew) {
+      accumulator.push({
+        insertOne: { document }
+      });
+    } else if (!utils.isEmptyObject(changes)) {
+      accumulator.push({
+        updateOne: {
+          filter: { _id: document._id },
```

Comment:
It would be better if this applied versioning and custom `where`. To do that, you'd need to do something like `document.$__version(_applyCustomWhere(document, document.$__where()), changes)`. Take a look at how `save()` does it.

Reply:
Great, will do.

I've been thinking about how this is going to play with middlewares, and how it's going to reflect on documents in case of failure. I am starting to think that we should change that into a helper:
```js
const writeOperations = Model.buildBulkWriteFromDocuments(documents);
await Model.bulkWrite(writeOperations);
```

I still haven't looked in the error object that's returned from bulkWrite failures, but unless it has a reference for the `_id` of the document that failed, we won't be able to know which documents to avoid marking as not-modified.

Reply:
@AbdelrahmanHafez is this ready to merge on your end? LGTM to me and I'd like to ship 5.12.0 in the next couple of weeks.

Reply:
@vkarpov15 
I've been giving this some thought, I'd like to expose two different functions:
A) `Model.buildWriteOperations(...)`: which takes an array of documents and returns an array of write operations, the implementation of this is quite simple.
B) `Model.bulkSave(...)`: This would utilize the first function and do a `bulkSave`, error-handling this one can have its nuances, we could fail mongoose-side validation errors early without sending any updates to the database, but for mongo-side errors, handling them can be a bit confusing to users.

How are we going to inform users as to which documents saves have succeeded and which ones failed?
How are we going to inform users of whether or not this is a mongo error or mongoose error?

I am starting to think that this will create more confusion than its worth, and for these reasons would rather just expose the first function as a nice utility that helps users bulkWrite and let them be responsible for how they use those operations.

I don't know if that's a thing in open-source, but I'm open to a pair-programming session where we brainstorm this. Let me know if that's something you'd be interested in.

I pushed the changes I have on mind.

As to the first utility, I've been running an experiment to compare the performance difference of using `updateMany` with the same `update` vs multiple `updateOne`s, and the results show around 2.5x performance improvement. The common update is a common use-case in our company (e.g. updating order `status` or some special identifier.). This may be a room for improvement in the first functionality. If you're interested in the details of the experiment, please check it out [here](https://stackoverflow.com/questions/65831219/mongodb-bulkwrite-multiple-updateone-vs-updatemany/65835186#65835186).

Reply:
I'd prefer to keep (A) internal, and expose (B). I think it's fine to just return a bulkWriteResult and let the client app figure out the bulkWriteResult as necessary. Like how `save()` does it: most of the time, users just look at the document to tell whether it was successfully saved or not.

The one big concern I have with this as implemented is `save()` hooks: right now `bulkSave()` doesn't trigger `save()` hooks, but perhaps it should. What do you think?